docs/rfc/0000-cli-infrastructure.adoc

```diff
@@ -0,0 +1,254 @@
+= RFC: RadiCLI
+:author: @FintanH
+:revdate: 2021-06-04
+:revremark: draft
+:toc:
+:toc-placement: preamble
+
+* Author: {author}
+* Date: {revdate}
+* Status: {revremark}
+* Community discussion: n/a
+* Tracking Issue: n/a
+
+== Motivation
+
+RFC-682 set the stage for defining the architecture of applications
+built on top of `<<radicle-link>>`. This RFC wishes to build on top of
+this specification and fill in the gaps around the CLI component and
+its organisation. As dogfooding our own solutions is on our long-term
+roadmap, this team wants to provide tooling that they will maintain
+and use themselves.
+
+== Overview
+
+This proposal will outline an binary called `rad`, what it will
+provide, and how it can be extended. Following this we provide details
+on the interaction -- or lack there of -- with the peer-to-peer
+daemon.
+
+We also provide a proposal of how any CLI tool will have
+access to any keys, as well as what needs to be completed for
+providing this.
+
+We close out with how the core components will be organised within
+`radicle-link` and `<<radicle-bins>>` as well as providing the first
+iteration of the core set of subcommands.
+
+== One `rad` to Rule Them All
+
+[quote]
+-------
+One ring to rule them all, one ring to find them,
+One ring to bring them all, and in the darkness bind them;
+-------
+-- J.R.R. Tolkein, Fellowship of the Ring
+
+The integral part of this proposal is that there will be one binary,
+`rad`, that will be the foundation for all other subcommands. It will
+define the global parameters and how to extend it.
+
+We will discuss these in the following two sections.
+
+== The Fellowship of the `rad`
+
+[quoute]
+--------
+“I can’t carry it for you, but I can carry you.” – Sam Gamgee
+--------
+-- J.R.R. Tolkien, The Return of the King
+
+To allow the ecosystem of commands to be extended by ourselves and
+others in the community it will be necessary to extend the `rad`
+command with further subcommands. The details of this are outlined in
+RFC-682#CLI, however, we will give a quick overview here for the sake
+of the reader.
+
+Taking inspiration from tools such as `git` and `cargo`, we propose
+that the `rad` binary can be extended by placing another
+binary on the users `$PATH` (or the Windows equivalent). This
+binary MUST follow the naming convention of `rad-*`. To reiterate
+here, these new commands MUST NOT override the core set outlined here
+(nor in further RFCs).
+
+== Global Parameters
+
+Many subcommands will likely want to share a set of options. These
+options, dubbed as "global parameters" here, will be a defined set --
+exported from `radicle-link` for re-use -- that the base `rad` command
+will always receive and pass onto the subcommand being called.
+
+Thus, all subcommands SHALL receive these global parameters as part of
```

Comment:
It would be good to detail _how exactly_ they receive those parameters. Builtin commands are obviously just function calls, but how is this handled for `execve`'d binaries? Does the `rad` wrapper parse (and validate) those options? Does it then just output them verbatim to the subcommand? Does it prefix them with `--global-` or something? Is order significant? Or does it use some other format like `CBOR` or `JSON`?

Reply:
I (hopefully) addressed this in https://github.com/radicle-dev/radicle-link/pull/698/commits/8a3d264dfbd3fd44cef3b27f29c3d7b07536a73c

Reply:
So if I’m not careful, I might accidentally expect a parameter which is already global 🤔 

Follow up question: is `println` the only contract for program output?

Reply:
> So if I’m not careful, I might accidentally expect a parameter that is already global

RTFM :grimacing: 

Ya, I was thinking what's the point of a global parameter if someone is just going to ignore it and use their own meaning of it. But given some more thought, it would be a Bad Thing:tm: if their behaviour was overridden by the globals. However, if `rad` prefixed them then it would need to pass both parameters, i.e. `--global-profile=abc` and `--profile=abc`, just in case there was a non-global version of the parameter defined. One would hope that wouldn't mess up a parser :crossed_fingers:

> Follow up question: is println the only contract for program output?

It hadn't crossed my mind. The other contract of the output would be the format. Is there anything else that I should think about here?

Reply:
> The other contract of the output would be the format.

Yeah that's the thing. It did cross my mind that @cloudhead (_just_ as an example) might fancy some ASCII spinners, emoji admonitions or whatever. Now, where would you override those? And how do you adhere to the style in an external subcommand?

Maybe this kind of thing is just not feasibly possible.

Reply:
Right, I think we should only be committing to plumbing commands here for now. I would say external commands can add plumbing on top of that and it's up to them how they mangle `stdout` :smile:

Reply:
Hmm, what if globals were passed as env vars to the sub-command, after being parsed?

Reply:
May, maybe it would still be good to define some kind of machine-readable output format at some point, which can be turned on with `—emoji-bufs` or something.